examples/vision/retinanet.py

```diff
@@ -0,0 +1,700 @@
+                )
+            )
+            head.add(keras.layers.ReLU())
+        head.add(
+            keras.layers.Conv2D(
+                output_filters,
+                3,
+                1,
+                padding="same",
+                kernel_initializer=kernel_init,
+                bias_initializer=bias_init,
+            )
+        )
+        return head
+
+    def call(self, image, training=False):
+        features = self.fpn(image, training=training)
+        N = tf.shape(image)[0]
+        cls_outputs = []
+        box_outputs = []
+        for feature in features:
+            box_outputs.append(tf.reshape(self.box_head(feature), [N, -1, 4]))
+            cls_outputs.append(
+                tf.reshape(self.cls_head(feature), [N, -1, self.num_classes])
+            )
+        cls_outputs = tf.concat(cls_outputs, axis=1)
+        box_outputs = tf.concat(box_outputs, axis=1)
+        return tf.concat([box_outputs, cls_outputs], axis=-1)
+
+
+"""
+## Implementing a custom layer to decode predictions
+"""
+
+
+class DecodePredictions(tf.keras.layers.Layer):
+    """Decodes RetinaNet Predictions"""
+
+    def __init__(
+        self,
+        num_classes=80,
+        confidence_threshold=0.05,
+        nms_iou_threshold=0.5,
+        soft_nms_sigma=0.4,
+        max_detections_per_class=100,
+        max_detections=100,
+        box_variance=[0.1, 0.1, 0.2, 0.2],
+        **kwargs
+    ):
+        super(DecodePredictions, self).__init__(**kwargs)
+        self.num_classes = num_classes
+        self.confidence_threshold = confidence_threshold
+        self.nms_iou_threshold = nms_iou_threshold
+        self.max_detections_per_class = max_detections_per_class
+        self.max_detections = max_detections
+
+        self._anchor_box = AnchorBox()
+        self._box_variance = tf.convert_to_tensor(
+            [0.1, 0.1, 0.2, 0.2], dtype=tf.float32
+        )
+
+    def _decode_box_predictions(self, anchor_boxes, box_predictions):
+        boxes = box_predictions * self._box_variance
+        boxes = tf.concat(
+            [
+                boxes[:, :, :2] * anchor_boxes[:, :, 2:] + anchor_boxes[:, :, :2],
+                tf.math.exp(boxes[:, :, 2:]) * anchor_boxes[:, :, 2:],
+            ],
+            axis=-1,
+        )
+        boxes_transformed = convert_to_corners(boxes)
+        return boxes_transformed
+
+    def call(self, images, predictions):
+        image_shape = tf.cast(tf.shape(images), dtype=tf.float32)
+        anchor_boxes = self._anchor_box.get_anchors(image_shape[1], image_shape[2])
+        box_predictions = predictions[:, :, :4]
+        cls_predictions = tf.nn.sigmoid(predictions[:, :, 4:])
+        boxes = self._decode_box_predictions(anchor_boxes[None, ...], box_predictions)
+        boxes = tf.expand_dims(boxes, axis=2)
+        boxes = tf.tile(boxes, [1, 1, self.num_classes, 1])
+
+        return tf.image.combined_non_max_suppression(
```

Comment:
I think it's using per class nms?

Reply:
Yes, this is per class nms. This gave a better mAP scores compared to simple nms. Should I  keep it configurable?

Reply:
I thought combined_nms is using all classes, which is different from per class?

Reply:
Sorry my bad, now I get what you are trying to say, I decided to go with `combined_non_max_suppression` and not per class nms is because it adding around 50 LOC, since the example is already very long I am not sure choose.